Fix bug in __get_attr__ where properties are inaccessible when pydantic_extra is populated

jmassucco17 opened the pull request:
## Change Summary
Fixes an issue where `__get_attr__` for `BaseModel` prioritizes `pydantic_extra` over general attributes resulting in an inability to access class properties when any extra fields are present.

The fix is to only prioritize `pydantic_extra` if the attribute name is actually present in the extra keys; otherwise, just go straight to attempting to access the attribute with `getattr()`

## Related issue number
fix #9862

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Viicos commented on pydantic/main.py:
```diff
@@ -878,7 +878,7 @@ def __getattr__(self, item: str) -> Any:
                 except AttributeError:
                     pydantic_extra = None
 
-                if pydantic_extra:
```
the `try..except` block isn't necessary with the given changes

Viicos commented:
Could you please share an example of what currently doesn't work? The added test already passes on `main`.

jmassucco17 commented:
> Could you please share an example of what currently doesn't work? The added test already passes on `main`.


Yup, thanks for pointing this out. I attempted to simplify down my real-world repro case but got too simple! This test should actually reproduce the issue; will update the PR later:

```py
class Model(pydantic.BaseModel):
    model_config = pydantic.ConfigDict(extra='allow')

    columns: List[str]

    @property
    def queried_columns(self) -> List[str]:
        return [c for c in self.columns if c.startswith('query')]


class SubModel(Model):
    pass


instance = SubModel.model_validate(
    dict(
        columns=['query_1', 'query_2', 'other'],
        test='extra_value',
    ))

# These have always worked
assert instance.queried_columns == ['query_1', 'query_2']
assert instance.__getattribute__('queried_columns') == ['query_1', 'query_2']

# This only works with the fix
assert instance.__getattr__('queried_columns') == ['query_1', 'query_2']

```

Viicos commented:
You shouldn't call `Model.__getattr__` manually. When doing `instance.queried_columns`, Python will first call `object.__getattribute__` which takes care of calling the `queried_columns` property's `__get__` method (because properties are non-data descriptors). Please refer to [this section](https://docs.python.org/3/howto/descriptor.html#invocation-from-an-instance) of the documentation to understand how `__getattribute__` works.

If it raises an `AttributeError`, Python will call our `BaseModel.__getattr__`, which we implemented _only_ to enable access to private/extra attributes/values. This way, doing `inst.test` works: `__getattribute__` raises an `AttributeError` because `test` is not in the instance's `__dict__`, and Python falls back to `BaseModel.__geattr__` which implements the logic to look up in `__pydantic_extra__`.

jmassucco17 commented:
> You shouldn't call `Model.__getattr__` manually. When doing `instance.queried_columns`, Python will first call `object.__getattribute__` which takes care of calling the `queried_columns` property's `__get__` method (because properties are non-data descriptors). Please refer to [this section](https://docs.python.org/3/howto/descriptor.html#invocation-from-an-instance) of the documentation to understand how `__getattribute__` works.
> 
> If it raises an `AttributeError`, Python will call our `BaseModel.__getattr__`, which we implemented _only_ to enable access to private/extra attributes/values. This way, doing `inst.test` works: `__getattribute__` raises an `AttributeError` because `test` is not in the instance's `__dict__`, and Python falls back to `BaseModel.__geattr__` which implements the logic to look up in `__pydantic_extra__`.


I'm only calling it manually in an attempt to illustrate the bug here. This issue is cropping up in real code that does not call __getattr__ directly, I'm simply trying to recreate it in a concise test case. This is the error I'm getting in my real code, as you can see all I'm doing is accessing the property `queried_columns` and this error shows up:
```
  File "org/grafana/tools/convert_to_sift.py", line 219, in <module>
    convert_to_sift(dashboard)
  File "org/grafana/tools/convert_to_sift.py", line 36, in convert_to_sift
    if not query.queried_columns:
  File "/home/jmassucco/devel/monorepo/dist/export/python/virtualenvs/python-default/3.8.16/lib/python3.8/site-packages/pydantic/main.py", line 850, in __getattr__
    raise AttributeError(f'{type(self).__name__!r} object has no attribute {item!r}') from exc
AttributeError: 'TimescaleQuery' object has no attribute 'queried_columns'
```

jmassucco17 commented:
I'm happy to share more of my code here to help get to the bottom of this, but I am quite confident there is a real bug somewhere here and I believe this was closed prematurely

Viicos commented:
Sorry, I got confused because your example shows that `instance.queried_columns` can be accessed. Do you happen to have an example that would lead to the `AttributeError` in your `convert_to_sift` function, when accessing `queried_columns`?

jmassucco17 commented:
> Sorry, I got confused because your example shows that `instance.queried_columns` can be accessed. Do you happen to have an example that would lead to the `AttributeError` in your `convert_to_sift` function, when accessing `queried_columns`?


Yeah my apologies - I am trying to work up a test case that can reproduce. My code that does reproduce is (a) quite complicated and (b) closed source, so it is a bit challenging to distill down my real failure case into something I can put in a test case. But I will work on it some more tonight and get back to you ASAP. Appreciate you reopening, looking forward to getting to the bottom of this

jmassucco17 commented:
@Viicos I constructed a test case based on my actual failure reproduction and pulled all the various dependencies into a single python file and dropped in a gist: https://gist.github.com/jmassucco17/b291ea6bc99676d571d13f1b7789414a

Running this reproduces the error I'm seeing. Interestingly, it totally depends on string contents of `raw_sql` field input to the model! Empty string doesn't reproduce, the real string (a bunch of SQL) does reproduce. But there are 1000s of other queries that I'm able to parse with no trouble so there's something specific about this one that is causing an issue...

jmassucco17 commented:
> @Viicos I constructed a test case based on my actual failure reproduction and pulled all the various dependencies into a single python file and dropped in a gist: https://gist.github.com/jmassucco17/b291ea6bc99676d571d13f1b7789414a
> 
> Running this reproduces the error I'm seeing. Interestingly, it totally depends on string contents of `raw_sql` field input to the model! Empty string doesn't reproduce, the real string (a bunch of SQL) does reproduce. But there are 1000s of other queries that I'm able to parse with no trouble so there's something specific about this one that is causing an issue...


I progressively went through and deleted text from the SQL and was able to get a surprisingly simple version to repro; it's surprising because the repro case is now so simple that I'm kind of shocked that I haven't run into this more... there must be something I'm not seeing

jmassucco17 commented:
OK I was able to narrow it down to an unguarded AttributeError from within the `sql_metadata` library that pydantic is claiming as its own.

The attribute error happens when we hit this line with the "bad" SQL: https://gist.github.com/jmassucco17/b291ea6bc99676d571d13f1b7789414a#file-repro-py-L189

I modified my code to directly access the `parsed_sql.tables` attribute of the instance and I get this error string:
```
Traceback (most recent call last):
  File "astranis/grafana/models/repro.py", line 283, in <module>
    example_instance.parsed_sql.tables
  File "/home/jmassucco/devel/monorepo/dist/export/python/virtualenvs/python-default/3.8.16/lib/python3.8/site-packages/sql_metadata/parser.py", line 345, in tables
    with_names = self.with_names
  File "/home/jmassucco/devel/monorepo/dist/export/python/virtualenvs/python-default/3.8.16/lib/python3.8/site-packages/sql_metadata/parser.py", line 451, in with_names
    token.next_token_not_comment.normalized
AttributeError: 'NoneType' object has no attribute 'normalized'
```

So I'm not sure if I would say this is a pydantic bug... but it does seem weird that pydantic is capturing and masking an `AttributeError` from another module like this. Do you think this is intended behavior, or something that could/should be handled more clearly?

jmassucco17 commented:
Closing this in favor of #10821 which has a more clear description of the issue and a more clear test case. I no longer feel confident that the fix I had proposed here is appropriate. Appreciate your help, @Viicos !
